src/bundle-manifest.ts

```diff
@@ -0,0 +1,160 @@
+/**
+ * A bundle is a grouping of files which serve the need of one or more
+ * entrypoint files.
+ */
+export interface IBundle {
+  entrypoints: Set<URL>;
+  files: Set<URL>;
+}
+
+/**
+ * A bundle strategy function is used to transform an array of bundles.
+ */
+export type IBundleStrategy = (bundles: IBundle[]) => IBundle[];
+
+/**
+ * A mapping of URLs to their full set of transitive dependencies, such that
+ * `a->b, b->c, b->d, e->f` would be represented `{a:[b,c,d], e:[f]}`
+ */
+export type IFileTransDepsIndex = Map<URL, Set<URL>>;
+
+/**
+ * Defining this URL type to make it clear which strings represent URLs.
+ */
+export type URL = string;
+
+/**
+ * Standard implementation of IBundle.
+ */
+export class Bundle implements IBundle {
+  public entrypoints: Set<URL>;
+  public files: Set<URL>;
+  constructor(entrypoints?: Set<URL>, files?: Set<URL>) {
+    this.entrypoints = entrypoints || new Set<URL>();
+    this.files = files || new Set<URL>();
+  }
+}
+
+/**
+ * Given an index of files and their dependencies, produce an array of bundles,
+ * where a bundle is defined for each set of dependencies.
+ *
+ * For example, a dependency index representing the graph:
+ *   `a->b, b->c, b->d, e->c, e->f`
+ *
+ * Would produce an array of three bundles:
+ *   `[a]->[b,d], [e]->[f], [a,e]->[c]`
+ */
+export function generateBundles(depsIndex: IFileTransDepsIndex): IBundle[] {
+  const bundles: IBundle[] = [];
+  const invertedIndex = invertFileTransDepsIndex(depsIndex);
+
+  for (let [dep, entrypoints] of invertedIndex.entries()) {
```

Comment:
Yeah this was the reason for Travis failures.  They were obscured by issues with type changes in parse5.  Thanks, guys- I was having a hard time identifying where the `[` token complaint was pointing to in the output.  *sigh*